src/boot/efi/stub.c

```diff
@@ -298,6 +328,71 @@ static void cmdline_append_and_measure_addons(
                 *cmdline_append = xasprintf("%ls %ls", tmp, copy);
 }
 
+static EFI_STATUS ucode_merge_and_measure_addons(
+                struct iovec initrds[static _INITRD_MAX],
+                struct iovec **ucode_addons,
+                size_t *n_ucode_addons,
+                int *sections_measured) {
+        EFI_STATUS err;
+
+        assert(initrds);
+        assert(ucode_addons);
+        assert(n_ucode_addons);
+
+        _cleanup_pages_ Pages all_ucode_combined_pages = {};
+        size_t all_ucode_combined_size = 0;
+
+        if (*n_ucode_addons == 0) {
+                return EFI_SUCCESS;
+        }
+
+        /* Invert ucode addons before measuring and merging, the kernel will take the first one. */
+        for (size_t i = 0; i < *n_ucode_addons / 2; i++) {
+                struct iovec *tmp = ucode_addons[i];
+                ucode_addons[i]= ucode_addons[*n_ucode_addons - 1 - i];
+                ucode_addons[*n_ucode_addons - 1 - i] = tmp;
+        }
+
+        /* Measure just the ucode addons, the uki-embedded one was already measured. */
+        FOREACH_ARRAY(addon, *ucode_addons, *n_ucode_addons) {
+                measure_section(POINTER_TO_PHYSICAL_ADDRESS(addon->iov_base), addon->iov_len, UNIFIED_SECTION_UCODE, sections_measured);
+        }
+
+        /* For merging, copy UKI-embedded addon and put it last in the list, so it has lowest pri. */
+        if (iovec_is_set(initrds + INITRD_UCODE)) {
+                *ucode_addons = xrealloc(*ucode_addons,
+                                *n_ucode_addons * sizeof(size_t),
+                                (*n_ucode_addons + 1)  * sizeof(size_t));
+                *ucode_addons[(*n_ucode_addons)++] = IOVEC_MAKE(
+                                xmemdup(initrds[INITRD_UCODE].iov_base, initrds[INITRD_UCODE].iov_len),
+                                initrds[INITRD_UCODE].iov_len);
+        }
+
+        err = combine_initrds(*ucode_addons, *n_ucode_addons, &all_ucode_combined_pages, &all_ucode_combined_size);
+        if (err != EFI_SUCCESS)
+                return err;
+
+        // copy the combined ucode initrd back into the initrds array to leave things in a consistent state for later cleanup
```

Comment:
coding style: please use `/* … */` style comments for explanatory stuff. We use `// …` style comments for fixme style stuff only.